packages/tpetra/core/example/BlockCrs/driver.cpp

```diff
@@ -0,0 +1,536 @@
+    else
+      exec_space::print_configuration(std::cout, false);
+  }
+
+  MeshDatabase mesh(comm,
+                    num_global_elements_i,
+                    num_global_elements_j,
+                    num_global_elements_k,
+                    num_procs_i,
+                    num_procs_j,
+                    num_procs_k);
+
+  const auto sb = mesh.getStructuredBlock();
+  const auto part = mesh.getStructuredBlockPart();
+
+  const auto num_owned_elements = mesh.getNumOwnedElements();
+  const auto num_remote_elements = mesh.getNumRemoteElements();
+  const auto num_owned_and_remote_elements = mesh.getNumElements();
+
+#if PRINT_VERBOSE
+  mesh.print(std::cout);
+#endif
+  
+  {
+    RCP<TimeMonitor> timerGlobal = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("X) Global")));
+    
+    // Build Tpetra Maps
+    // -----------------
+    
+    const global_ordinal_type TpetraComputeGlobalNumElements 
+      = Teuchos::OrdinalTraits<global_ordinal_type>::invalid();  
+    
+    // - all internal views are allocated on device; mirror as mesh database is constructed on host  
+    const auto mesh_gids_host = mesh.getElementGlobalIDs();
+    const auto mesh_gids = Kokkos::create_mirror_view(typename exec_space::memory_space(), mesh.getElementGlobalIDs());
+    Kokkos::deep_copy(mesh_gids, mesh_gids_host);
+
+    // for convenience, separate the access to owned and remote gids
+    const auto owned_gids  = Kokkos::subview(mesh_gids, local_ordinal_range_type(                 0, num_owned_elements));
+    const auto remote_gids = Kokkos::subview(mesh_gids, local_ordinal_range_type(num_owned_elements, num_owned_and_remote_elements));
+    
+    RCP<const map_type> row_map = rcp(new map_type(TpetraComputeGlobalNumElements, owned_gids, 0, comm));
+    RCP<const map_type> col_map = rcp(new map_type(TpetraComputeGlobalNumElements, mesh_gids, 0, comm));
+    
+#if PRINT_VERBOSE
+    row_map->describe(*out);
+    col_map->describe(*out);
+#endif
+    
+    // Graph Construction
+    // ------------------
+    // local graph is constructed on device space
+    typedef tpetra_crs_graph_type::local_graph_type local_graph_type;
+    typedef local_graph_type::row_map_type::non_const_type rowptr_view_type;
+    typedef typename local_graph_type::entries_type colidx_view_type;
+
+    rowptr_view_type rowptr;
+    colidx_view_type colidx;
+    {
+      RCP<TimeMonitor> timerLocalGraphConstruction 
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("0) LocalGraphConstruction")));
+      
+      rowptr = rowptr_view_type("rowptr", num_owned_elements + 1);
+
+      local_ordinal_range_type owned_range_i, owned_range_j, owned_range_k;
+      local_ordinal_range_type remote_range_i, remote_range_j, remote_range_k;
+
+      part.getOwnedRange(owned_range_i, owned_range_j, owned_range_k);
+      part.getRemoteRange(sb, remote_range_i, remote_range_j, remote_range_k);
+      
+      // count # of nonzeros per row
```

Comment:
Tpetra has a function for this, called Tpetra::Details::computeOffsetsFromCounts:

https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_computeOffsets.hpp

Reply:
We wouldn't expect users (e.g., writers of BlockCRS assembly in an application) to call functions in Tpetra::Details, would we?  If not, then we shouldn't call them from our benchmark examples.

Reply:
This is reasonable.  I think it would still make sense to break this out into a separate function in the example, though.  Examples should be easy to read and follow general good programming practices.

Reply:
If I use the API, it calls parallel loops twice (one to parallel compute # of nonzeros per row and the other for parallel scan - and the parallel scan is two pass algorithm).